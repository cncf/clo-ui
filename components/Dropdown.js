module.exports = require('../dist/cjs/components/Dropdown.js');
