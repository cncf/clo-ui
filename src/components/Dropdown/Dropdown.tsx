import classNames from 'classnames';
import isUndefined from 'lodash/isUndefined';
import React, { useEffect, useRef, useState } from 'react';

import { useOutsideClick } from '../../hooks/useOutsideClick';
import styles from './Dropdown.module.css';

export interface IDropdownProps {
  label: string;
  btnContent: React.ReactNode;
  children: React.ReactElement<{
    closeDropdown?: () => void;
    isVisibleDropdown?: boolean;
  }>;
  btnClassName?: string;
  dropdownClassName?: string;
  onClose?: () => void;
  disabled?: boolean;
}

export const Dropdown: React.FC<IDropdownProps> = (props: IDropdownProps) => {
  const [visibleDropdown, setVisibleDropdown] = useState(false);
  const ref = useRef<HTMLDivElement | null>(null);
  useOutsideClick([ref], visibleDropdown, () => setVisibleDropdown(false));

  const closeDropdown = () => {
    setVisibleDropdown(false);
  };

  useEffect(() => {
    if (!visibleDropdown && !isUndefined(props.onClose)) {
      props.onClose();
    }
  }, [visibleDropdown]);

  return (
    <div ref={ref} className="position-relative">
      <button
        className={`btn btn-md btn-link text-white rounded-0 lh-1 ${styles.btn} ${props.btnClassName}`}
        type="button"
        onClick={() => setVisibleDropdown(!visibleDropdown)}
        aria-label={`${props.label} button`}
        aria-expanded={visibleDropdown}
        disabled={props.disabled}
      >
        {props.btnContent}
      </button>

      <div
        role="menu"
        className={classNames('dropdown-menu rounded-0', styles.dropdown, props.dropdownClassName, {
          show: visibleDropdown,
        })}
      >
        <div className={`dropdown-arrow ${styles.arrow}`} />
        {React.isValidElement(props.children)
          ? props.children.type === React.Fragment
            ? (() => {
                const fragment = props.children as React.ReactElement<{ children?: React.ReactNode }>;
                return (
                  <>
                    {React.Children.map(fragment.props.children, (child) =>
                      React.isValidElement(child)
                        ? React.cloneElement(child as React.ReactElement<Record<string, unknown>>, {
                            closeDropdown: closeDropdown,
                            isVisibleDropdown: visibleDropdown,
                          })
                        : child
                    )}
                  </>
                );
              })()
            : React.cloneElement(props.children, { closeDropdown: closeDropdown, isVisibleDropdown: visibleDropdown })
          : props.children}
      </div>
    </div>
  );
};
