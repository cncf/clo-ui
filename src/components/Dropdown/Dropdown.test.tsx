import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Dropdown } from './Dropdown';

const defaultProps = {
  label: 'Label',
  btnContent: 'Button content',
};

describe('Dropdown', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <Dropdown {...defaultProps}>
        <>Dropdown content</>
      </Dropdown>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders proper content', () => {
    render(
      <Dropdown {...defaultProps}>
        <>Dropdown content</>
      </Dropdown>
    );

    expect(screen.getByRole('button')).toBeInTheDocument();
    expect(screen.getByRole('button')).toHaveTextContent('Button content');
    expect(screen.getByLabelText('Label button')).toBeInTheDocument();
    expect(screen.getByRole('menu')).toBeInTheDocument();
    expect(screen.getByRole('menu')).not.toHaveClass('show');
  });

  it('opens dropdown', async () => {
    render(
      <Dropdown {...defaultProps}>
        <>Dropdown content</>
      </Dropdown>
    );

    expect(screen.getByRole('menu')).not.toHaveClass('show');

    const button = screen.getByRole('button');
    await userEvent.click(button);

    expect(screen.getByRole('menu')).toHaveClass('show');
  });
});
