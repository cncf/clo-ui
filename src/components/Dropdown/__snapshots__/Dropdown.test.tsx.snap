// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="position-relative"
  >
    <button
      aria-expanded="false"
      aria-label="Label button"
      class="btn btn-md btn-link text-white rounded-0 lh-1 btn undefined"
      type="button"
    >
      Button content
    </button>
    <div
      class="dropdown-menu rounded-0 dropdown"
      role="menu"
    >
      <div
        class="dropdown-arrow arrow"
      />
      Dropdown content
    </div>
  </div>
</DocumentFragment>
`;
